### PR TITLE
Subscriptions: Polish contract

### PR DESCRIPTION
**Pull request by facuspagnuolo**
**MERGE AFTER #137**

### CODE REVIEW

File: contracts/subscriptions/CourtSubscriptions.sol [hunk 1]
```diff
@@ -121,15 +131,15 @@ contract CourtSubscriptions is IsContract, ISubscriptions, TimeHelpers {
...
-        require(_periodDuration > 0, ERROR_ZERO_PERIOD_DURATION);
+        require(_periodDuration > 0, ERROR_PERIOD_DURATION_ZERO);
 
         owner = _owner;
         jurorsRegistry = _jurorsRegistry;
         periodDuration = _periodDuration;
         _setFeeToken(_feeToken);
         _setFeeAmount(_feeAmount);
         _setPrePaymentPeriods(_prePaymentPeriods);
-        latePaymentPenaltyPct = _latePaymentPenaltyPct;
```
Comment (bingen) [inline contracts/subscriptions/CourtSubscriptions.sol]: I don't get this. What's the point of converting a simple assignment into a function? I would say this is quite atomic, simple and readable. Are we supposed to convert all assignments into functions?
Besides, as far as I know Solidity doesn't have `inline` functions (modifiers a part).

File: contracts/subscriptions/CourtSubscriptions.sol [hunk 2]
```diff
@@ -221,7 +239,9 @@ contract CourtSubscriptions is IsContract, ISubscriptions, TimeHelpers {
     function setFeeToken(ERC20 _feeToken, uint256 _feeAmount) external onlyGovernor {
         // `setFeeToken` transfers governor's accumulated fees, so must be executed first
         _setFeeToken(_feeToken);
-        _setFeeAmount(_feeAmount);
+        if (currentFeeAmount != _feeAmount) {
```
Comment (bingen) [inline contracts/subscriptions/CourtSubscriptions.sol]: :+1:

File: contracts/subscriptions/CourtSubscriptions.sol [hunk 3]
```diff
@@ -414,43 +421,71 @@ contract CourtSubscriptions is IsContract, ISubscriptions, TimeHelpers {
...
     * @dev Internal function to set a new amount for the subscription fees
-    * @param _feeAmount New amount of fees to be paid for each subscription period
+    * @param _newFeeAmount New amount of fees to be paid for each subscription period
     */
-    function _setFeeAmount(uint256 _feeAmount) internal {
-        require(_feeAmount > 0, ERROR_ZERO_FEE);
-        currentFeeAmount = _feeAmount;
+    function _setFeeAmount(uint256 _newFeeAmount) internal {
+        uint256 previousFeeAmount = currentFeeAmount;
+        require(previousFeeAmount != _newFeeAmount, ERROR_INVALID_FEE_AMOUNT);
```
Comment (bingen) [inline contracts/subscriptions/CourtSubscriptions.sol]: This sounds like too much overhead to me, unnecessary complexity and gas waste. What's the problem if somebody tries to set the amount to the same one?
Same applies to other setters below.

File: contracts/subscriptions/CourtSubscriptions.sol [hunk 4]
```diff
@@ -469,7 +504,7 @@ contract CourtSubscriptions is IsContract, ISubscriptions, TimeHelpers {
     */
     function _getPeriodStartTermId(uint256 _periodId) internal view returns (uint64) {
         // Periods are measured in Court terms. Since Court terms are represented in `uint64`, we are safe to use `uint64` for period ids too.
-        return START_TERM_ID + uint64(_periodId) * periodDuration;
+        return START_TERM_ID + uint64(_periodId * periodDuration);
```
Comment (bingen) [inline contracts/subscriptions/CourtSubscriptions.sol]: Why this change? `periodDuration` is already an `uint64`.

File: contracts/subscriptions/CourtSubscriptions.sol (hunk 1 above)
Comment (facuspagnuolo) [reply contracts/subscriptions/CourtSubscriptions.sol]: Please, review the setter first

File: contracts/subscriptions/CourtSubscriptions.sol (hunk 3 above)
Comment (facuspagnuolo) [reply contracts/subscriptions/CourtSubscriptions.sol]: Emitting duplicated events, I'd always prioritize usability vs gas consumption, even more if we are talking about non-daily-basis user functionality like this function

File: contracts/subscriptions/CourtSubscriptions.sol (hunk 4 above)
Comment (facuspagnuolo) [reply contracts/subscriptions/CourtSubscriptions.sol]: True, not sure what I meant here, maybe it was a typo 🤷‍♂

File: contracts/subscriptions/CourtSubscriptions.sol (hunk 3 above)
Comment (bingen) [reply contracts/subscriptions/CourtSubscriptions.sol]: Hm, ok, sounds reasonable.

File: contracts/subscriptions/CourtSubscriptions.sol (hunk 1 above)
Comment (bingen) [reply contracts/subscriptions/CourtSubscriptions.sol]: Good point, I was assuming it would do the same. Still, I don't think we need all that extra stuff. We could discuss about the event (do we  need it on `init`?), but right now it would be impossible to init the contract wit a `latePaymentPenaltyPct = 0`, right? I don't think this is what we want.

Comment (facuspagnuolo) [reply contracts/subscriptions/CourtSubscriptions.sol]: True, updated the internal setters to avoid checking diffs

Comment (bingen) [reply contracts/subscriptions/CourtSubscriptions.sol]: I see you removed the check but left the function and the event. I still don't think we need the event on initialization (and therefore the function). Not sure, as this is more a front-end issue, but we are not doing this in aragon-apps. See some examples:
- Finance:
https://github.com/aragon/aragon-apps/blob/master/apps/finance/contracts/Finance.sol#L177
https://github.com/aragon/aragon-apps/blob/master/apps/finance/contracts/Finance.sol#L311

- Voting:
https://github.com/aragon/aragon-apps/blob/master/apps/voting/contracts/Voting.sol#L86
https://github.com/aragon/aragon-apps/blob/master/apps/voting/contracts/Voting.sol#L103
https://github.com/aragon/aragon-apps/blob/master/apps/voting/contracts/Voting.sol#L117

File: contracts/subscriptions/CourtSubscriptions.sol (hunk 3 above)
Comment (bingen) [reply contracts/subscriptions/CourtSubscriptions.sol]: Altough btw we don't do this on aragon-apps either

https://github.com/aragon/aragon-apps/blob/master/apps/voting/contracts/Voting.sol#L95
https://github.com/aragon/aragon-apps/blob/master/apps/voting/contracts/Voting.sol#L110
https://github.com/aragon/aragon-apps/blob/master/apps/finance/contracts/Finance.sol#L304

We could bring it up there. Pinging @sohkai to hear his opinion about it.

File: contracts/subscriptions/CourtSubscriptions.sol [hunk 5]
```diff
@@ -514,31 +543,53 @@ contract CourtSubscriptions is IsContract, ISubscriptions, TimeHelpers {
...
+    * @param _subscriber Subscriber querying the delayed periods of
+    * @param _currentPeriodId Identification number of the current period
+    * @return Number of overdue payments for the requested subscriber
+    *
+    *    subs      last           cur       new
+    *   +----+----+----+----+----+----+----+----+
+    *                  <---------><------------->
+    *                    delayed      regular
+    *                  <------------------------>
+    *                          _periods
```
Comment (bingen) [inline contracts/subscriptions/CourtSubscriptions.sol]: Quick reminder: either move it back to `payFees` function or change this `_periods` variable that doesn't belong to here.

File: contracts/subscriptions/CourtSubscriptions.sol (hunk 3 above)
Comment (bingen) [reply contracts/subscriptions/CourtSubscriptions.sol]: I've been thinking more about it and I'm again not convinced with this change. This can be easily prevented in the front-end. I remember once @izqui telling me that our job is not to make front-end life easy :wink: and I guess that's why we don't so these things in aragon-apps. I guess we should keep that philosophy.

Comment (facuspagnuolo) [reply contracts/subscriptions/CourtSubscriptions.sol]: > Altough btw we don't do this on aragon-apps either

Nice, a few more things to fix then :)

> This can be easily prevented in the front-end.

Everything can be prevented by the front-end

Seriously, no strong feelings, emitting "falsey" events is not a good practice IMO. This is sth I'd expect from an auditor's feedback for example. However, I don't want to waste time on something non-critical, I'll roll back this behavior.

Comment (bingen) [reply contracts/subscriptions/CourtSubscriptions.sol]: > Everything can be prevented by the front-end

Haha, fair enough, but you know what I meant right? That as you said this is not critical.

> emitting "falsey" events is not a good practice IMO

Yes, as I said I see your points... although prevent "falsey" events would be as simple as renaming them from `...Changed` to `...Set` :stuck_out_tongue_winking_eye: 

> I'll roll back this behavior

Sounds good, although maybe we can wait for @izqui or @sohkai to confirm that this is a common and intended practice of ours. Or maybe we can just cherry-pick this if we decide we want it.

Comment (bingen) [reply contracts/subscriptions/CourtSubscriptions.sol]: Just for the record, I want to clarify this a little bit more for further discussion (as "not critical" sounds too vague)

>>> I'd always prioritize usability vs gas consumption
>> This can be easily prevented in the front-end.
> Everything can be prevented by the front-end

The way I see we have been working so far is: security (and gas consumption) must be taken care of in smart contracts, usability belongs to front-end (of course there are exceptions, like bundling two transactions in one)

File: contracts/subscriptions/CourtSubscriptions.sol [hunk 6]
```diff
@@ -514,31 +535,51 @@ contract CourtSubscriptions is IsContract, ISubscriptions, TimeHelpers {
...
+    /**
+    * @dev Internal function to tell the number of overdue payments for a given subscriber
+    * @param _subscriber Subscriber querying the delayed periods of
+    * @param _currentPeriodId Identification number of the current period
+    * @return Number of overdue payments for the requested subscriber
+    *
+    *    subs      last           cur       new
+    *   +----+----+----+----+----+----+----+----+
+    *                  <---------><------------->
+    *                    delayed      regular
```
Comment (bingen) [inline contracts/subscriptions/CourtSubscriptions.sol]: To be honest, I don't think this makes much sense the way it is here. This was intended to explain the whole picture, how the paid periods (this line has now been removed) were the sum of `delayed` plus `regular` and how it related to timeline. So the `regular` part doesn't apply here.
I'd rather remove it.
